source/language/openpulse.rst

```diff
@@ -0,0 +1,650 @@
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   kernel gaussian(complex[size] amp, length l, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   kernel sech(complex[size] amp, length l, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // square_width is the width of the square waveform component
+   // sigma is the standard deviation of waveform
+   kernel gaussian_square(complex[size] amp, length l, length square_width, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   // beta is the Y correction amplitude, see the DRAG paper
+   kernel drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   kernel constant(complex[size] amp, length l) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   // frequency is the frequency of the waveform
+   // phase is the phase of the waveform
+   kernel sine(complex[size] amp, length  l, float[size] frequency, angle[size] phase) -> waveform;
+
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms
+
+- ``mix(waveform wf1, waveform wf2)`` -> waveform - Mix two input waveforms to produce a new waveform.
+  This is equivalent to the product signal :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
+- ``sum(waveform wf1, waveform wf2)`` -> waveform - Sum two input waveforms to produce a new waveform.
+- ``piecewise(waveform wf0, waveform wf1)`` -> waveform - Output waveform.
+- ``offset(waveform wf, complex amount)`` -> waveform - Offset the input waveform by an amount.
+- ``scale(waveform wf, complex factor)`` -> waveform - Scale the input waveform by a factor.
+- ``conj(waveform wf) -> waveform`` - Conjugate the input waveform.
+- ``re(waveform wf) -> waveform`` - Real component of input waveform.
+- ``im(waveform wf) -> waveform`` - Imaginary component of input waveform.
+- ``abs(waveform wf) -> waveform`` - Transform waveform as norm of input. waveform
+- ``phase_shift(waveform wf, angle ang) -> waveform`` - Signal with relative phase, ang.
+
+Play instruction
+----------------
+
+Waveforms are scheduled using the ``play`` instruction. These instructions may
+only appear inside a ``defcal`` block and have three required parameters:
+
+- the channel on which to play the pulse
+- a value of type ``waveform`` representing the waveform envelope
+- the frame to use for the pulse
+
+.. code-block:: javascript
+
+  play(channel chan, waveform wfm, frame fr)
```

Comment:
Potential modification: Define a virtual channel type `vxchannel` (or other name). Tie channels to frames in the form of virtual channels, eg., `vxchannel chan = virtual(channel, frame)` and do `play(vxchannel vc_chan, waveform wfm)`. Virtual channel tracks time based on the frame time.

Reply:
That would be probably be the most often used case, but I wonder if the minor convenience of not having to write both the channel and frame in a `play` instruction worth the complexity (however small) of adding yet another type?

Reply:
That is a fair point. One benefit of the virtual channel approach is that timing is tracked in a linear way and never needs to be done across channels...

Reply:
We could also tie the virtual channel to qubits rather than doing it in `setframe/getframe`.

Reply:
This issue was discussed with the group. It was decided that we leave things as they are for the following reasons:
- Issue of updating frames: when we update the channel, does it update the frame or do we have to do this separately?
- Frame should contain timing truth
- Geometric gates are less clear w/ virtual channels